keps/sig-instrumentation/20191115-structured-logging.md

```diff
@@ -0,0 +1,556 @@
+
+#### Logger initialization
+
+Logger implementation initialization should be done at startup of component and passed down the stack as an interface. Aside of initialization of logger, there should not be any dependency on specific implementation of logger Interface.
+
+#### Storing Logger and metadata in Context
+
+We would like to propose that logger interface and all metadata collected trough call stack to be stored in Golang Context object. This way creating one canonical way to pass contextual information across the call stack. Some pros of this approach:
+* Standard method in Golang to pass scoped variables across API boundaries and between processes.
+* No need for additional function argument for passing logging API through call stack
+* Context is used similarly by many frameworks including for instance OpenCensus tracing API.
+* Accessing metadata will be independent from Logging and could be reused by other functionalities.
+
+Access to context variables should be done though predefined functions. For that we propose to a new module to kubernetes component base called `k8s.io/component-base/kontext`. This new module will include all utility functions for storing and retrieving logger and metadata needed by it.
+
+Throughout call stack new metadata can be added to context that can be later used by logger. For example reconciliation loop can reference the reconcilided object so that all logs will be referencing it. All metadata access should be done through functions in `k8s.io/component-base/kontext` module. Adding reference to kubernetes object (pod, service, node) should be really straightforward.
+
+#### Example
+
+```go
+import "k8s.io/component-base/kontext"
+
+func UpdateStatus(ctx context.Context, pods []*v1.Pod, newStatus bool) {
+	for _, pod := range pods {
+		podCtx := kontext.WithObject(ctx, pod)
+		updatePodStatus(podCtx, pod, newStatus)
+	}
+}
+
+func updatePodStatus(ctx context.Context, pod *v1.Pod, newStatus bool) {
+  logger := kontext.GetLogger(ctx).WithName("probe")
+  logger.V(6).Info("Updating ready status", "ready", newStatus)
+  [...]
+  logger.V(4).Info("Ready status updated", "ready", newStatus)
+}
+```
+
+### New logging formats
+
+Introduction of new logging API should be paired with new output format that can properly utilize structured logging. In order to simplify log ingestion, but still maintain similar developer experience we would like to propose two formats. First one for production use and second one for component development.
+
+#### Json format
+
+We would like to propose a JSON as a future default format in Kubernetes for both production and development use. Some pros of using JSON:
+* Broadly adopted by logging libraries with very efficient implementations (zap, zerolog).
+* Out of the box support by many logging backends (Elasticsearch, Stackdriver, BigQuery, Splunk)
+* Easily parsable and transformable
+* Existing tools for ad-hoc analysis (jq)
+
+Example:
```

Comment:
please add examples of both external (versioned) and internal (hub) API objects.